MMDevice/MMDevice.h

```diff
@@ -205,6 +208,104 @@ namespace MM {
    };
 
 
+   ////////////////////////////
+   ///// Standard properties
+   ////////////////////////////
+   
+   //// Helper functions/macros for compile time and runtime checking of standard properties
+
+   // Map from device types to standard properties
+   inline std::unordered_map<MM::DeviceType, std::vector<StandardProperty>>& GetDeviceTypeStandardPropertiesMap() {
+      static std::unordered_map<MM::DeviceType, std::vector<StandardProperty>> devicePropertiesMap;
+      return devicePropertiesMap;
+   }
+
+   // Register a standard property with its valid device types
+   inline void RegisterStandardProperty(const StandardProperty& prop, std::initializer_list<MM::DeviceType> deviceTypes) {
+      for (MM::DeviceType deviceType : deviceTypes) {
+         GetDeviceTypeStandardPropertiesMap()[deviceType].push_back(prop);
+      }
+   }
+
+   // Check if a property is valid for a specific device type
+   inline bool IsPropertyValidForDeviceType(const StandardProperty& prop, MM::DeviceType deviceType) {
+      auto it = GetDeviceTypeStandardPropertiesMap().find(deviceType);
+      if (it == GetDeviceTypeStandardPropertiesMap().end()) {
+         return false;
+      }
+      
+      const auto& validProperties = it->second;
+      return std::find(validProperties.begin(), validProperties.end(), prop) != validProperties.end();
+   }
+
+   // The below code is a way to enable compile time checking of which 
+   // standard properties are are valid for which device types. This enables
+   // methods for setting these standard properties defined once 
+   // in CDeviceBase and then conditionally enabled in child classes based on
+   // the LINK_STANDARD_PROP_TO_DEVICE_TYPE macro below
+   // In addition to making this link, a dedicated method for the standard property
+   // should be created in CDeviceBase.
+   template <MM::DeviceType DeviceType, const MM::StandardProperty& PropRef>
+   struct IsStandardPropertyValid {
+      static const bool value = false; // Default to false
+   };
+
+   // The top struct enables compile time checking of standard property
+   // creation methods (ie that you can call the CreateXXXXStandardProperty
+   // in a device type that doesn't support it)
+   // The bottom part adds the standard property to a runtime registry
+   // which enables higher level code the query the standard properties
+   // associated with a particular device type and check that everything
+   // is correct (ie all required properties are implemented after
+   // device initialization)
+   #define LINK_STANDARD_PROP_TO_DEVICE_TYPE(DeviceType, PropertyRef) \
```

Comment:
Since macros are not confined to a namespace, best to name with a prefix. Here, perhaps `MM_INTERNAL_` so that it's clear that user device code should not use this macro.